### PR TITLE
Extend Fedora to support kernel updates on chromebooks

### PR DESCRIPTION
**Pull request by dorindabassey**
Added a kernel-install script that creates an intramfs and FIT Image (including the ramdisk)
then the script flashes the kernel to the right partition in order to support kernel updates on the chromebook

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

### CODE REVIEW

File: 96-chromebook.install [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+if [[ -d "$BOOT_DIR_ABS" ]]; then
+    INITRD="initrd"
+else
+    BOOT_DIR_ABS="/boot"
+    INITRD="initramfs-${KERNEL_VERSION}.img"
+fi
+
+case "$COMMAND" in
+    add)
+        echo "Install required packages"
```
Comment (martinezjavier) [inline 96-chromebook.install]: I don't think other kernel-install scripts print an output ?

File: 96-chromebook.install [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        dnf install lz4 vboot-utils uboot-tools -y
```
Comment (martinezjavier) [inline 96-chromebook.install]: I would prefer if this script instead was shipped let's say as a part of vboot-utils? or maybe a vboot-utils-kernel-install sub-package and pulled these as dependencies ?

File: 96-chromebook.install [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if [[ -f "$BOOT_DIR_ABS/$INITRD" ]]; then
+            rm -f -- "$BOOT_DIR_ABS/$INITRD"
+        fi
+        dracut --force -v --force-drivers "ulpi usb-storage phy-qcom-usb-hs-28nm \
+        phy-qcom-usb-ss ocmem dwc3 dwc3-of-simple dwc3-pci ehci-platform xhci-plat-hcd \
+        i2c-qcom-geni i2c-qup icc-osm-l3 qcom-spmi-pmic phy-qcom-qmp phy-qcom-qusb2 \
+        phy-qcom-usb-hs qcom_aoss qcom-apcs-ipc-mailbox llcc-qcom nvmem_qfprom smem \
+        smp2p dwc3-qcom \
+        " "$BOOT_DIR_ABS/$INITRD" --kver "$KERNEL_VERSION" --kmoddir /lib/modules/"${KERNEL_VERSION}"
```
Comment (martinezjavier) [inline 96-chromebook.install]: I don't think all this has to be done here? this be handled by the dracut kernel-install plugin (`/usr/lib/kernel/install.d/50-dracut.install`) already.

File: 96-chromebook.install [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        kernel="${KERNEL_IMAGE##*/}-${KERNEL_VERSION}.lz4"
+        compression="lz4"
```
Comment (martinezjavier) [inline 96-chromebook.install]: is this compression type required by Coreboot ?

File: 96-chromebook.install [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        # Compress image
+        rm -f "$BOOT_DIR_ABS/${KERNEL_IMAGE##*/}-${KERNEL_VERSION}.lz4" || true
+        lz4 "$BOOT_DIR_ABS/${KERNEL_IMAGE##*/}-${KERNEL_VERSION}" "$BOOT_DIR_ABS/${KERNEL_IMAGE##*/}-${KERNEL_VERSION}.lz4"
+        dtbs=" \
+            -b /boot/dtb/qcom/sc7180-trogdor-coachz-r3.dtb \
+            -b /boot/dtb/qcom/sc7180-trogdor-lazor-r3-kb.dtb \
+            -b /boot/dtb/rockchip/rk3399-gru-kevin.dtb\
+            -b /boot/dtb/rockchip/rk3399-gru-scarlet-inx.dtb \
+            "
```
Comment (martinezjavier) [inline 96-chromebook.install]: Instead of hardcoding this, maybe we can pick all the files in `/boot/dtb/*` ?

File: 96-chromebook.install [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        mkimage -D "-I dts -O dtb -p 2048" -i "$BOOT_DIR_ABS/$INITRD" \
+                -f auto -A arm64 -O linux -T kernel -C $compression -a 0 \
+                -d /boot/$kernel $dtbs \
+                kernel.itb
+
+	bootloader="boot_params"
+	echo "root=PARTUUID=%U/PARTNROFF=1 rootwait rw" > boot_params
```
Comment (martinezjavier) [inline 96-chromebook.install]: No, instead just use the cmdline in `/proc/cmdline`. Or better:

```
if [[ -f /etc/kernel/cmdline ]]; then
    read -r -d '' -a BOOT_OPTIONS < /etc/kernel/cmdline
elif [[ -f /usr/lib/kernel/cmdline ]]; then
    read -r -d '' -a BOOT_OPTIONS < /usr/lib/kernel/cmdline
else
    declare -a BOOT_OPTIONS

    read -r -d '' -a line < /proc/cmdline
    for i in "${line[@]}"; do
        [[ "${i#initrd=*}" != "$i" ]] && continue
        [[ "${i#BOOT_IMAGE=*}" != "$i" ]] && continue
        BOOT_OPTIONS+=("$i")
    done
fi
```

like is done by `/usr/lib/kernel/install.d/90-loaderentry.install`.

File: 96-chromebook.install [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	vbutil_kernel --pack /boot/kernel.vboot \
+		               --keyblock /usr/share/vboot/devkeys/kernel.keyblock \
+		               --signprivate /usr/share/vboot/devkeys/kernel_data_key.vbprivk \
+		               --version 1 --config boot_params \
+		               --bootloader $bootloader \
+		               --vmlinuz kernel.itb \
+		               --arch arm
+	dd if=/boot/kernel.vboot of=/dev/sda1 bs=4M status=progress
```
Comment (martinezjavier) [inline 96-chromebook.install]: we shouldn't hardcode `/dev/sda1` but instead figure out what is the kernel partition using `cgpt` or something like that.

File: 96-chromebook.install (hunk 2 above)
Comment (martinezjavier) [reply 96-chromebook.install]: just to make it clear, I don't think that the script should install packages. In any case if you want to ship this with `chromebook-setup.sh`, then these packages should be added as dependencies in the script itself using `ensure_command`.

File: 96-chromebook.install (hunk 6 above)
Comment (dorindabassey) [reply 96-chromebook.install]: noted

File: 96-chromebook.install (hunk 2 above)
Comment (dorindabassey) [reply 96-chromebook.install:18]: like do you mean submitting a patch to include the 96-chromebook.install script in vboot-utils ustream?

File: 96-chromebook.install (hunk 3 above)
Comment (dorindabassey) [reply 96-chromebook.install]: this was the list of required modules provided by enric for the initramfs, so this section will regenerate an initramfs to suit the chromebook.

File: 96-chromebook.install (hunk 2 above)
Comment (martinezjavier) [reply 96-chromebook.install]: yes. But I would ship it as a separate sub-package since it may be that someone wants to have the vboot-utils tools in their host machine for cross-compile and then ship to a target device. So a `vboot-kernel-install` or something sub-package would make sense.

That way you won't need to manually install the dependencies for this since it would be handled by the package manager.

Comment (dorindabassey) [reply 96-chromebook.install]: oh I get what you mean now, thanks.

File: 96-chromebook.install (hunk 3 above)
Comment (martinezjavier) [reply 96-chromebook.install]: yes, that makes sense for creating an initial initrd but that's not needed when updating a kernel package. Since dracut would run on the target machine in that case and would figure out what modules need to be included.

File: 96-chromebook.install (hunk 4 above)
Comment (dorindabassey) [reply 96-chromebook.install]: this is the compression method already being used by coreboot for the Initial kernel in the chromebook

Comment (martinezjavier) [reply 96-chromebook.install]: I see. Maybe worth adding a comment then?

File: 96-chromebook.install (hunk 5 above)
Comment (dorindabassey) [reply 96-chromebook.install]: yeah, this make sense but i don't think the fedora kernel generates all the device tree in /boot/dtb/* @eballetbo what d'you think.

Comment (martinezjavier) [reply 96-chromebook.install]: but you are already relying on these DTB existing there... I just think that having those hardcoded here won't scale.

File: 96-chromebook.install (hunk 7 above)
Comment (dorindabassey) [reply 96-chromebook.install]: Alright, thanks. will fix that

File: 96-chromebook.install (hunk 3 above)
Comment (dorindabassey) [reply 96-chromebook.install]: Okay, I see. Thanks

File: 96-chromebook.install (hunk 5 above)
Comment (dorindabassey) [reply 96-chromebook.install]: I think temporarily we can use these know dtb that the fedora kernel can generate then we can open a new ticket to figure out why the other existing dtb doesn't work with the FIT image. what d'you think?

File: chromebook-setup.sh [hunk 8]
```diff
@@ -737,8 +741,8 @@ cmd_setup_fedora_kernel()
 
     # Extract and copy the kernel packages to the rootfs
     mkdir ./tmpdir && cd ./tmpdir
-    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idmv
-    rpm2cpio ../kernel-modules-$kernel_version.rpm | cpio -idmv
+    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idm
```
Comment (eballetbo) [inline chromebook-setup.sh]: This is an unrelated change. Please remove

File: chromebook-setup.sh [hunk 9]
```diff
@@ -737,8 +741,8 @@ cmd_setup_fedora_kernel()
...
+    rpm2cpio ../kernel-modules-$kernel_version.rpm | cpio -idm
```
Comment (eballetbo) [inline chromebook-setup.sh]: Same here. Please remove that change

File: chromebook-setup.sh [hunk 10]
```diff
@@ -845,7 +849,7 @@ ensure_command realpath realpath
 ensure_command sgdisk gdisk
 ensure_command lz4 lz4
 ensure_command mkfs.ext4 e2fsprogs
-ensure_command mkimage u-boot-tools
+ensure_command mkimage uboot-tools
```
Comment (eballetbo) [inline chromebook-setup.sh]: Hmm, so debian names this package as u-boot-tools whether fedora names it as uboot-tools. Not sure how e need to handle these cases but in any case I think it should be a separate patch. So I'd prefer do not change it in this PR.

**Review by eballetbo — CHANGES_REQUESTED** [review]
Thanks for the PR. I requested some minor comments, let me know if you're fine.

File: chromebook-setup.sh (hunk 10 above)
Comment (dorindabassey) [reply chromebook-setup.sh]: okay, i see. fedora names it otherwise. I'd create a separate patch to fix that then

File: 96-chromebook.install [hunk 11]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/bash
```
Comment (eballetbo) [inline 96-chromebook.install]: Can you prepend de commit message title with `chromebook-setup:`

`chromebook-setup: Extend Fedora to support kernel updates on Chromebooks`

Just to follow the same pattern used in all other commits.

File: chromebook-setup.sh [hunk 12]
```diff
@@ -742,6 +742,10 @@ cmd_setup_fedora_rootfs()
     -e '/home/s/btrfs/ext4/' \
     -e 's/subvol=home,compress=zstd:1/defaults/' ./tmpdir/root/etc/fstab
 
+    # echo "Inserting Kernel Update script"
+    sudo chmod 777 96-chromebook.install
+    sudo cp 96-chromebook.install ./tmpdir/root/usr/lib/kernel/install.d/
```
Comment (eballetbo) [inline chromebook-setup.sh]: Instead of putting this file inside the source path, can we create a new directory called scripts and put the file there?

**Review by eballetbo — CHANGES_REQUESTED** [review]
Hi Dorinda, two minor comments, more cosmetic than problematic, other than that the patch looks good to me.

File: 96-chromebook.install (hunk 11 above)
Comment (dorindabassey) [reply 96-chromebook.install]: ok will do thanks

File: chromebook-setup.sh (hunk 12 above)
Comment (dorindabassey) [reply chromebook-setup.sh]: cool, make sense

**Comment by eballetbo** [comment]
Merged. Many thanks for contributing.
